Add Signal channel

tehCh0nG opened the pull request:
Apologies for the mass commits. I wasn't able to get them to squash.

Commit:
Add Signal channel

Revert "Add Discord channel" (#14)

Update README.md

chore: update instructions

chore: update readme

docs: fork repo correction

fix: update documentation

Add files via upload

Create src folder

Create Exceptions folder

Delete Exceptions

Create Delete

Create config folder

Create delete

Upload Signal files

Upload file

Create delete

Add files via upload

Add files via upload

Delete Delete

Delete delete

Delete delete

Delete test

Delete signal-notification-channel.php

Add files via upload

Delete delete

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Create src folder

tehCh0nG committed:
Create src folder

tehCh0nG committed:
Create Exceptions folder

tehCh0nG committed:
Delete Exceptions

tehCh0nG committed:
Create Delete

tehCh0nG committed:
Create Delete

tehCh0nG committed:
Create config folder

tehCh0nG committed:
Create config folder

tehCh0nG committed:
Create delete

tehCh0nG committed:
Create delete

tehCh0nG committed:
Upload Signal files

tehCh0nG committed:
Upload file

tehCh0nG committed:
Create delete

tehCh0nG committed:
Create delete

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Delete Delete

tehCh0nG committed:
Delete Delete

tehCh0nG committed:
Delete delete

tehCh0nG committed:
Delete delete

tehCh0nG committed:
Delete delete

tehCh0nG committed:
Delete delete

tehCh0nG committed:
Delete test

tehCh0nG committed:
Delete test

tehCh0nG committed:
Delete signal-notification-channel.php

tehCh0nG committed:
Delete signal-notification-channel.php

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Delete delete

tehCh0nG committed:
Delete delete

Commit:
Merge remote-tracking branch 'refs/remotes/origin/master'

Commit:
Merge branch 'master' of github.com:tehCh0nG/channels

atymic commented on composer.lock:
```diff
@@ -0,0 +1,4921 @@
+{
```
No need to commit this, can be gitignored.

tehCh0nG replied:
Added to gitignore.

atymic commented on src/Exceptions/CouldNotSendNotification.php:
```diff
@@ -0,0 +1,11 @@
+{
+    public static function serviceRespondedWithAnError($response)
+    {
+        return new static("Unable to send message. Please ensure recipient is registered with Signal. Symfony output was: `{$response}['$symfonyerror']`");
```
where does `symfonyerror` come from?

atymic replied:
Could we use `sprintf` please

tehCh0nG replied:
Changed to `sprintf` and `symfonyerror` to `result`.

atymic commented on src/Exceptions/CouldNotSendNotification.php:
```diff
@@ -0,0 +1,11 @@
+
+class CouldNotSendNotification extends \Exception
+{
+    public static function serviceRespondedWithAnError($response)
```
Could we type this.

tehCh0nG replied:
Typed in latest revision.

atymic commented on src/SignalChannel.php:
```diff
@@ -0,0 +1,59 @@
+
+        //Run signal-cli via Symfony Process.
+        $result = new Process(
+            ['/path/to/signal-cli/bin', '--username',$username,'send','--message',$message,$recipient],
```
Should this not be a config option?

tehCh0nG replied:
Done.

atymic commented on src/SignalChannel.php:
```diff
@@ -0,0 +1,59 @@
+
+        $message = Arr::flatten($message);
+
+        $username = $message[0];
```
Can we not flatten this please?

tehCh0nG replied:
Changed to collection.

atymic commented on src/SignalChannel.php:
```diff
@@ -0,0 +1,59 @@
+            ['/path/to/signal-cli/bin', '--username',$username,'send','--message',$message,$recipient],
+            //Pass JAVA_HOME to Symfony so signal-cli can run.
+            null,
+            ['JAVA_HOME' => '/path/to/java']
```
Also

tehCh0nG replied:
Done

atymic commented on src/SignalChannel.php:
```diff
@@ -0,0 +1,59 @@
+
+        $result->run(); //run once
+
+        while ($result->isRunning()) {
```
This is not required as process->run is blocking
https://symfony.com/doc/current/components/process.html#usage

tehCh0nG replied:
Removed

atymic commented on src/SignalChannel.php:
```diff
@@ -0,0 +1,59 @@
+         }
+
+        if (!$result->isSuccessful()) {
+            $symfonyerror = new ProcessFailedException($result);
```
Wrong var passed here.

atymic replied:
Shouldn't we pass a string, not another exception object here?

atymic commented on src/SignalMessage.php:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @return static
+     */
+    public static function create($message = '')
```
types please (return types as well)

tehCh0nG replied:
Changed

atymic commented on src/tests/MessageTest.php:
```diff
@@ -0,0 +1,35 @@
+
+class MessageTest extends TestCase
+{
+  // @var NotificationChannels\Signal\SignalMessage
```
```suggestion
  /** @var NotificationChannels\Signal\SignalMessage */
```

atymic replied:
Still wrong

atymic commented on src/tests/MessageTest.php:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  public function sets_username()
+  {
```
Could we get some more in depth tests please?

tehCh0nG replied:
Added test for message.

tehCh0nG committed:
Update src/tests/MessageTest.php

Fix code comment

Co-authored-by: atymic <atymicq@gmail.com>

tehCh0nG committed:
Add files via upload

Add config options, function types, additional tests.
Removed flattening.
Add .gitignore.

tehCh0nG committed:
Delete CouldNotSendNotification.php

tehCh0nG committed:
Delete MessageTest.php

tehCh0nG committed:
Delete SignalChannel.php

tehCh0nG committed:
Delete SignalMessage.php

tehCh0nG committed:
Delete signal-notification-channel.php

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Delete signal-notification-channel.php

tehCh0nG committed:
Add files via upload

tehCh0nG committed:
Delete composer.lock

Not needed

atymic commented on src/Exceptions/CouldNotSendNotification.php:
```diff
@@ -0,0 +1,11 @@
+{
+    public static function serviceRespondedWithAnError(string $response)
+    {
+        return sprintf("Unable to send message. Please ensure recipient is registered with Signal. Symfony output was: `{$response}['$result']`");
```
This won't work still, and please use sprintf

atymic commented on src/Signal.php:
```diff
@@ -0,0 +1,15 @@
+      *  @var string phone number for recipient;
+      * ex: +12345556789, +442012345678
+      **/
+        $this->sender => env('SIGNAL_USERNAME', false);
```
please don't use envs directly, will break with config caching

atymic commented on src/SignalChannel.php:
```diff
@@ -0,0 +1,51 @@
+
+        //Run signal-cli via Symfony Process.
+        $result = new Process(
+            [config('signal-notification-channel.signal_cli'), '--username',config('signal-notification-channel.username'),'send','--message',$message,$recipient],
```
Could you split this onto multiple lines?

atymic commented on src/SignalChannel.php:
```diff
@@ -0,0 +1,51 @@
+
+        $result->run();
+
+         if (!$result->isSuccessful()) {
```
indenting here is a bit wonky

atymic commented on src/SignalMessage.php:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @return $this
+     */
+    public function message($message)
```
still missing type hints on all the other methods here. Also, indenting is wonky as well.

tehCh0nG committed:
Add files via upload

Fix indenting, remove env usage, fix spacing

tehCh0nG committed:
Add files via upload

Added additional tests

tehCh0nG committed:
Add files via upload

Fix sprintf usage

tehCh0nG committed:
Update SignalMessage.php

Add type hints

atymic commented:
@tehCh0nG still happy to do this? If so i will re-review. 
Thanks
